In [32]:
import pandas as pd
import numpy as np
from sklearn.metrics import confusion_matrix

### Read the set of labels into a dataframe:

In [33]:
labels = pd.read_csv('labels.csv')

In [34]:
labels.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 149 entries, 0 to 148
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   perfect_labeler  149 non-null    object 
 1   radiologist      149 non-null    object 
 2   algorithm        149 non-null    float64
dtypes: float64(1), object(2)
memory usage: 3.6+ KB


In [35]:
labels.head()

perfect_labeler radiologist  algorithm
0          cancer      cancer       0.99
1          cancer      cancer       0.94
2          cancer      cancer       0.73
3          cancer      cancer       0.82
4          cancer      cancer       0.98

In [36]:
labels['perfect_labeler'].value_counts()

benign    120
cancer     29
Name: perfect_labeler, dtype: int64

### Start with assessing the radiologist's performance:
* Assess the _accuracy_ of the radiologist by just looking at the percent of cases that they correctly labeled
* Next, look at the true positive and true negative rates of the radiologist by generating a _confusion matrix_ 

In [37]:
radiologist_accuracy = sum(labels.perfect_labeler == labels.radiologist)/len(labels)

In [38]:
radiologist_accuracy

0.8993288590604027

In [39]:
confusion_matrix(labels.perfect_labeler.values,labels.radiologist.values,labels=["cancer","benign"])

array([[ 25,   4],
       [ 11, 109]])

### Now look at the algorithm's performance compared to the perfect labeler:
* Since the algorithm doesn't create a binary label, it instead returns a _probability_ of cancer, choose a probability cut-off to use for the algorithm's labeling of cancer vs. bening. _(Hint: 0.5 is a reasonable starting place)_
* Start with assessing _accuracy_ again here
* Generate a confusion matrix

In [40]:
labels = labels.replace('cancer', 1).replace('benign', 0)
labels.head()

perfect_labeler  radiologist  algorithm
0                1            1       0.99
1                1            1       0.94
2                1            1       0.73
3                1            1       0.82
4                1            1       0.98

In [41]:
threshold = (labels.algorithm > 0.5)
print(threshold)
print(type(threshold))

0       True
1       True
2       True
3       True
4       True
       ...  
144    False
145     True
146     True
147    False
148    False
Name: algorithm, Length: 149, dtype: bool
<class 'pandas.core.series.Series'>


In [42]:
confusion_matrix(labels.perfect_labeler.values, threshold.values, labels=[1,0])

array([[ 21,   8],
       [  8, 112]])

### Confusion Matrix Interpretation

1. We can clearly see that among 30 positive cases as per perfect labeler, our ML algorithm was able to correctly identity 21.
2. Sesitivity = (21/30) = 0.7
3. We can clearly see that among 120 negative cases as per perfect labeler, our ML algorithm was able to correclt identify 112.
3. Specificity = (112/120) = 0.933

What happens now if you change the threshold cut-off for your algorithm's classification to 0.4? What if you raise it to 0.6? How do accuracy, fp, fn, tp, and tn change?

In [43]:
# changing threshold to 0.4

threshold1 = (labels.algorithm > 0.4)
confusion_matrix(labels.perfect_labeler.values, threshold1.values, labels=[1,0])

array([[ 25,   4],
       [ 16, 104]])

### Confusion Matrix Interpretation

1. We can clearly see that among 31 positive cases as per perfect labeler, our ML algorithm was able to correctly identity 25.
2. Sesitivity = (25/31) = 0.806
3. We can clearly see that among 108 negative cases as per perfect labeler, our ML algorithm was able to correclt identify 104.
3. Specificity = (104/108) = 0.963

In [44]:
# Changing threshold to 0.6

threshold2 = (labels.algorithm > 0.6)
confusion_matrix(labels.perfect_labeler.values, threshold2.values, labels=[1,0])

array([[ 20,   9],
       [  5, 115]])

### Confusion Matrix Interpretation

1. We can clearly see that among 25 positive cases as per perfect labeler, our ML algorithm was able to correctly identity 20.
2. Sesitivity = (20/25) = 0.8
3. We can clearly see that among 108 negative cases as per perfect labeler, our ML algorithm was able to correclt identify 104.
3. Specificity = (115/124) = 0.927

### Finally, let's compare our algorithm to the radiologist
* A "perfect labeler" might not exist in the real world, and in fact, if often does not
* In AI for medical imaging, using a radiologist's labels as our "true" label is often the standard of practice, and algorithm performance is judged in both an academic setting as well as in the regulated industry landscape based on performance against an expert human

* Repeat the steps above using a set threshold for your algorithm (again, 0.5 is perfectly reasonable) but now computing accuracy, tp, tn, fp, fn against the radiologist. 

In [45]:
threshold3 = (labels.algorithm > 0.5)
confusion_matrix(labels.radiologist.values, threshold.values, labels=[1,0])

array([[ 23,  13],
       [  6, 107]])

### Confusion Matrix Interpretation

1. We can clearly see that among 29 positive cases as per radiologist, our ML algorithm was able to correctly identity 23.
2. Sesitivity = (23/29) = 0.793
3. We can clearly see that among 120 negative cases as per radiologist, our ML algorithm was able to correclt identify 107.
3. Specificity = (107/120) = 0.891

## Reflection: 
* In the above exercise you assess performances of a human as well as of an algorithm against a 'perfect labeler' and also against each other. 
* Does accuracy seem like the appropriate statistic to use when evaluating these labels? Why or why not? 
* In what clinical settings does it seem more or less acceptable to have a high level of FNs? FPs? 
* How did changing the threshold on the algorithm performance change the different performance statistics? 
* How did your opinion of the algorithm's performance change when you started comparing it to a radiologist instead of the perfect labeler? What does this mean for a real-world scenario when a perfect labeler doesn't exist, and we only have a radiologist's read to base our performance on? 

### Analysis

Given that there are so many fewer cancer cases than benign cases, accuracy would _not_ be a good statistic to use. A human or an algorithm could label _all_ of the cancer cases as benign and still achieve an accuracy of over 80%. 

Higher levels of false negatives aren't great ever in clinical settings, but they have less of an impact on the patient if we _know_ that there will be a second reader (i.e. our algorithm reads an image first, and then the label is confirmed by a radiologist). This also would only be appropriate if there wasn't a time-sensitive aspect to making the diagnosis. It seems more acceptable to have a high level of false positives in a situation as we saw in the previous exercise, where our algorithm was being used to prioritize emergency reads, in which case we would want to be somewhat liberal. 

Changing the threshold on the algorithm performance changed the FP and FN rates, one at the expense of the other. 

The algorithm's true positive rate increased when using the same threshold and comparing it to the human instead of the perfect labeler. This means that without the _true_ ground truth of diagnoses in an image, we may never be able to 100% accurately assess our algorithm, and its results may be inflated based on the quality of the radiologist labels that we use in comparing our outputs. 